src/server/src/main/java/io/cassandrareaper/core/StreamSession.java

```diff
@@ -0,0 +1,184 @@
+    sb.append("]");
+    return sb.toString();
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Initializes a new stream within the session.
+   *
+   * <p>This method is called to handle a notification about new stream being established.
+   *
+   * @param host originating the notification
+   * @param sessionInfo the payload received from Cassandra with the JMX notification
+   * @param timestamp of the JMX notification
+   * @return itself with the new stream added
+   */
+  public StreamSession addStream(String host, SessionInfo sessionInfo, long timestamp) {
+    Stream newStream = Stream.newStream(host, sessionInfo, timestamp);
+    streams.putIfAbsent(newStream.streamId(), newStream);
+    return this;
+  }
+
+  /**
+   * Updates a stream within the session based on the update received from Cassandra.
+   *
+   * <p>The method assumes the stream has been previously initialized when {@link #addStream(String, SessionInfo, long)}
+   * was called. If it wasn't, the method does nothing.
+   *
+   * <p>If the notification comes with old timestamp, then no update happens.
+   *
+   * <p>This method is called every time Cassandra sends a stream update, which happens quite a lot.
+   *
+   * @param host originating the notification
+   * @param progressInfo the payload received from Cassandra with the JMX notification
+   * @param timeStamp of the JMX notification
+   * @return itself with the stream updated
+   */
+  public StreamSession updateStream(String host, ProgressInfo progressInfo, long timeStamp) {
+    String streamId = Stream.streamId(host, progressInfo.peer.toString());
+
+    streams.computeIfPresent(streamId, (id, oldStream) -> {
+      // process the payload only if the timestamp is newer
+      if (oldStream.getLastUpdated() < timeStamp) {
+        if (progressInfo.direction.equals(ProgressInfo.Direction.OUT)) {
+          oldStream.setSizeSent(progressInfo);
+        }
+        if (progressInfo.direction.equals(ProgressInfo.Direction.IN)) {
+          oldStream.setSizeReceived(progressInfo);
+        }
+        oldStream.setLastUpdated(timeStamp);
+      }
+      return oldStream;
+    });
+    return this;
+  }
+
+  /**
+   * Updates the stream to mark it as complete after Cassandra sends the related message.
+   *
+   * <p>The method assumes the stream has been initialized with {@link #addStream(String, SessionInfo, long)}.
+   * If it wasn't, then the method does nothing.
+   *
+   * <p>If the notification comes with old timestamps, then no update happens.
+   *
+   * @param host originating the notification
+   * @param peer the second node participating in the stream
+   * @param success of the streaming session between the two nodes
+   * @param timeStamp of the JMX notification
+   * @return itself with stream finalized
+   */
+  public StreamSession finalizeStream(String host, String peer, Boolean success, long timeStamp) {
```

Comment:
`addStream()`, `updateStream()`  and `finalizeStream()` should be taken out of this class to comply with the design of POJO classes. Maybe those should be part of the StreamManager class instead, reworked by taking a StreamSession object as additional argument.